### ISSUE TITLE
withLatestFrom doesn't actually provide latest value

### ISSUE THREAD

**Issue by dimalexeev** [ReactiveX/RxSwift#1121]
```
let subItem = Variable<SubItem?>(nil)
let item = Variable<Item?>(nil)
let result = Variable<Result?>(nil)

subItem.asObservable()
    .withLatestFrom(item.asObservable()) { ($0, $1) }
    .debug()
    .map { <some transformation here> }
    .bindTo(details)
    .addDisposableTo(bag)

item.asObservable()
  .map { <some transformation here> }
  .bindTo(subItem)
  .addDisposableTo(bag)
```

This code causes the following actions

* if item Variable emits next event with the first non nil value
* and because of that subitem Variable emits next event with the first non nil value
* then debug() printing shows that next event contains (non nil value, nil value)

In other words **.withLatestFrom(item.asObservable()) can't provide actual latest next event's non nil value, but provide the previous one's nil value!**

Possible solutions.

1. Use item.value instead of withLatestFrom(item.asObservable())
2. Create a constant

```
let itemAsSharedObservable = item.asObservable().shareReplayLatestWhileConnected()
```

And use it everywhere instead of item.asObservable()

**Could you please clarify a little bit**

1. What is really happening here?
2. What is the best practice to avoid this issue?

**RxSwift/RxCocoa/RxBlocking/RxTest version/commit**

3.1.0

**Platform/Environment**

  - [x] iOS
  - [ ] macOS
  - [ ] tvOS
  - [ ] watchOS
  - [ ] playgrounds

**How easy is to reproduce?**

  - [x] easy, 100% repro
  - [ ] sometimes, 10%-100%
  - [ ] hard, 2% - 10%
  - [ ] extremely hard, %0 - 2%

**Xcode version**:

```
8.2.1
```

**Installation method**:
  - [x] CocoaPods
  - [ ] Carthage
  - [ ] Git submodules

**I have multiple versions of Xcode installed**:
  - [ ] yes (which ones)
  - [x] no

**Level of RxSwift knowledge**:
  - [ ] just starting
  - [x] I have a small code base
  - [ ] I have a significant code base

**Comment by kzaher** [comment]
Hi @dimalexeev ,

you have created a system with strong feedback loop because you've used a lot of variables. Using Rx won't help you much if you don't setup your architecture in a unidirectional way and you use a lot of `Subjects`.

`withLatestFrom` operator also assumes that first observable is event (doesn't emit initial value immediately), and second is state like (emits first value as soon as possible).

I can't give you a better answer than, don't create feedback loops/cycles by using `Variable`s and don't use `withLatestFrom` with two `Variable`s.

**Comment by dimalexeev** [comment]
But why do you name this issue strong loop issue, if "withLatestFrom" should just ask for latest value? Nothing more as far as I understand documentation. It is not, for example, "combineLatest" which calls observer on every combined observable item arrival.

**Comment by kzaher** [comment]
Hi @dimalexeev ,

Both `.withLatestFrom(item.asObservable())` and `item.asObservable()
  .map { <some transformation here> }...` register observers on `item`. 
The order of sending events to those observers is arbitrary, there is no order guarantee.

`withLatestFrom` assumes that the two observable sequences it uses as arguments don't fire simultaneously like in your case (that they have different sources).

If they fire simultaneous that means that the problem could be reformulated by just using a single sequence without with latest from operator, like it can in your case.

The bottom line is, just use one variable typed `(SubItem?, Item?, Result?)` and when you need some part use `map` operator to select the part you need.
